### ISSUE TITLE
TagUI-Python vs TagUI - summary of differences and roadmap

### ISSUE THREAD

**Issue by ck81** [tebelorg/TagUI-Python#47]
### Background
Have spent the last few weeks trying to convert one complicated script I’ve written using TagUI into TagUI-Python to print 3 different reports across two corporate websites. The printing of the report is not that straightforward as it involves some lookup of values before you can print the report. To minimize the duplication of codes, I stored those commonly used functions in separate files and call these functions using the `tagui` command.  As a result, the final app contains more than 50 files!

### Result
Pleased to say that I have successfully ported the entire TagUI script into TagUI-Python! The app runs perfectly without any glitch! As I can now combine multiple functions into Python modules, the number of files also drastically reduced form 50+ files to only 3 main files + 4 Python modules. 

The only commands that I’ve found to be different between TagUI and TagUI-Python are the two I’ve highlighted earlier:
* write #44
* Run #45 

Other than the above, as TagUI-Python runs on top of TagUI, there’s almost no changes required. All that’s required is the translation of the syntax e.g.

* read
TagUI: `read //path/to/selector to var_name`
TagUI-Python: `var_name = t.read('//path/to/selector')`
* type
TagUI: `type //path/to/selector as string_to_type`
TagUI-Python: `t.type('//path/to/selector', 'string_to_type')`

**Comment by ck81** [comment]
### TagUI-Python vs TagUI - some highlights
On surface, there are no much difference between developing a script using TagUI as oppose to TagUI-Python.  However, let me highlight here some advantages of coding using TagUI-Python. 

#### 1. Combining multiple functions into one Python modules
TagUI-Python is a standard Python package that runs entirely within the standard Python environment.  This means that you can use all the native python data structures such as lists, dictionaries and modules, etc. With Python modules, I was able to combine and consolidate similar functions from 50 over files into just 4 python modules!

#### 2. Python classes and subclasses
With TagUI-Python, I was able to use classes, subclasses, inheritance and overriding of methods to encapsulates common functionalities of the 3 reports into 1 main report class and 3 sub-classes.  This makes the maintenance of codes so much easier because any changes in the future (e.g. if there's a change in the layout of the website), I only need to make changes in one place. Any bug fix or new feature enhancements will also be directly applied to all 3 reports.  

#### 3. Adding your own customized TagUI Functions
Using Python functions , you can now easily create your own customized TagUI functions. For example, one command I really like and use a lot is the `hover` command. It gives a clear visual indication of which element the robot is working on now and allows the user to follow clearly what’s going on. Imagine you need to grab information row by row from a table of 100 rows. I always issue a hover first before each read. In this way, the user can visually see which row the robot is processing now. 

In TagUI, you have to do the following:
```
hover selector
read selector to str
```

Although it seems ok, but imagine there’s now a change in the website layout, and there and 30 of these hover's and read's, you need to update each pair for all the 30 pairs. 

In Python, I can create a new function as follows:
```python
def hover_and_click(selector):
	t.hover(selector)
	t.click(selector)
```

Now all I need to do is to call:
```python
hover_and_click(selector) 
```
for each row of the table

Another example is a common use case I often encounter.  In a website, some of the elements are not present every time. So I often need to test if the element is present first before I can read the element, otherwise TagUI will throw an error if you try to read an element that is not there. In TagUI, the code looks like this:

```javascript
mobile_number = ''
if present(selector)
{
    read selector to mobile_number
}
```

With TagUI-Python, I created a new function
```python
def read_if_present(selector):
    str = ''
    if t.present(selector):
        str = t.read(selector)
    return str
```

I can now easily read an optional element in a web page with the one-liner:
```
mobile_number = read_if_present(selector)
```

#### 4. Using standard python packages
As TagUI-Python runs within the standard Python environment, you have access to all the standard Python packages. Some of the packages that I’ve used in my app includes:
* import re - for using regular expressions in string search and replace
* import os - for checking if file exists and files operations such as moving of files
* import datetime - for getting current date and time, getting the date for tomorrow, next Monday, etc.
* import time - for calculating elapsed time
* import pandas - for reading and write .csv file

**Comment by ck81** [comment]
### Conclusion
Through this exercise, I have found that TagUI-Python can do everything I can achieve using TagUI, with a lot more additional advantages as highlighted above. In terms of speed and performance, I also notice no visible difference.

So yes, I will be using TagUI-Python moving forward.

I still see value in TagUI.  The one unique feature for TagUI is the `LIVE` command - especially useful when you want to test a website interactively.

**Comment by kensoh** [comment]
(tagging as feature label as above posts share on features of TagUI for Python)

Thanks CK for your thorough review of TagUI for Python vs TagUI! Added comments below -

- for live mode, you can type `python` from prompt to enter Python's live mode to test TagUI for Python's commands interactively; you can also do that via Jupyter notebook environment
- write() is intentionally different as the target users are different, thus behaviour is adjusted to use best practice for programming languages (does not auto-add line break) - #44
- run() is fixed in v1.10 to work in Windows, available with `pip install tagui --upgrade`

TagUI for Python is a version of TagUI that only exposes the UI interaction features. TagUI is more like a complete framework that also has features like [repositories](https://github.com/kelaberetiv/TagUI#repositories) to store element identifiers, and [datatables](https://github.com/kelaberetiv/TagUI#datatables) to run batch automation. Also, the human language features become irrelevant in Python, eg the [21 native languages](https://github.com/kelaberetiv/TagUI#native-languages), as well as writing [conditions statements](https://github.com/kelaberetiv/TagUI#conditions-examples) in human language.

However, what is lacking is made up by the list of advantages you shared above, mainly due to advantages of being based fully on a programming language, plus immense benefits from the Python ecosystem (lots of packages, documentations, tutorials, etc).

And of course, for users doing machine-learning work, now they can use RPA functions as part of their work (where their work is the focus and RPA is the supporting role), versus using TagUI and calling Python integration (where RPA is the focus and machine-learning capability is the supporting role). There are repetitive work in data acquisition, cleaning, preparation and model training that can be assisted and made easier with RPA. TagUI for Python tries to play that supporting role well.

Lastly, by taking the step to create this Python package, I would like to signal to developers in other programming languages that TagUI's API is mature and stable enough to build other new stuff base on it. The eventual goal for this project is to make it easy for developers in other programming languages to port this simple architecture and code to their favourite programming language, and making RPA directly accessible to users of that language.

The package has been installed over 1000 times so far. And there has been only 1 bug-fix in the past month (https://github.com/tebelorg/TagUI-Python/commits/master). There should be a surge of new users in Sep. When that happens, it will be opportunity for me to address any new bugs discovered or missing features, and bring this package from Beta to Production status.
